secure_software_development_fundamentals.md

```diff
@@ -2411,17 +2431,121 @@ Of course, like any technique, if you use it wrongly then it won’t be secure.
+
+For more information on using parameterized queries
+in stored procedures, see your library's documentation, the
+[OWASP Query Parameterization Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Query_Parameterization_Cheat_Sheet.html), and the
+[OWASP SQL Injection Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html)
+
+##### When Parameterized Interfaces Won't Work
+
+In some situations parameterized statements (including
+prepared statements) will *not* work.
+Many parameterized interface APIs only allow replacing SQL values, so
+they do not allow varying information such as the names of tables, the names
+of columns, or the sort order direction.
+
+In those cases you may need to create a query by concatenating data.
+In those cases, make *sure* you carefully validate the data (using an allowlist)
+so only specific and safe values are allowed.
+Often the allowlist is a short list of permitted values, or at most
+a simple expression that only allows ASCII letters and digits.
+A risk with this approach is that if the validation
+is ever skipped (e.g., after some code change),
+the system may be become extremely vulnerable.
+
+The [OWASP SQL Injection Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html)
+has a nice suggestion: if you *must* do this concatenation,
+make the user input *different* from what you will actually use, and
+in the program map the user input to the values you will use.
+If user input doesn't have a valid mapping, reject the input.
+This increases the probability that a change to the program
+will not result in unvalidated input being concatenated into a query,
+or that the problem will be detected before shipping.
+
+##### Other Approaches
+
+Many programs use object-relational mapping (ORM). This is just a technique to automatically convert data in a relational database into an object in an object-oriented programming language and back; lots of libraries and frameworks will do this for you. This is fine, as long as the ORM is implemented using parameterized statements or something equivalent to them. In practice, any good ORM implementation will do so. So if you are using a respected ORM, you are already doing this. That said, it is common in systems that use ORMs to occasionally need to use SQL queries directly… and when you do, use parameterized statements or prepared statements.
+
+Some applications use a "query builder" library to build commands (queries) programmatically through a sequence of calls instead of embedding a command string. Some ORMs include a query builder system. Again, a well-implemented query builder will use parameterized statements or similar internally. So if you use a query builder, use that is implemented using parameterized statements and provide untrusted data as separate parameters however in the query builder requires for them to be correctly escaped.
```

Comment:
```suggestion
Some applications use a "query builder" library to build commands (queries) programmatically through a sequence of calls instead of embedding a command string. Some ORMs include a query builder system. Again, a well-implemented query builder will use parameterized statements or similar internally. So if you use a query builder, use one that is implemented using parameterized statements and provides untrusted data as separate parameters however in the query builder requires for them to be correctly escaped.
```

I also think this needs a grammar check ("provides untrusted data as separate parameters however in the query builder requires for them to be correctly escaped."), not sure what you meant to say